e2e: wait for pods with gomega, II

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

gomega.Eventually provides better progress reports: instead of filling up the
log with one-line messages that are not enough to to understand the
current state, it integrates with Gingko's progress reporting (SIGUSR1,
--poll-progress-after) and then dumps the same complete failure message as
after a timeout. That makes it possible to understand why progress isn't
getting made without having to wait for the timeout.
    
The code also is generic enough that it becomes more reusable:
- error handling (= retry after "not found", delay) can be used
  for all kind of objects
- multiple checks can be combined with gomega.And

#### Which issue(s) this PR fixes:

Related-to #https://github.com/kubernetes/kubernetes/issues/106575

#### Special notes for your reviewer:

These commits were previously merged in https://github.com/kubernetes/kubernetes/pull/113298, then reverted in https://github.com/kubernetes/kubernetes/pull/115543 because of regressions caused by the WaitForPodToDisappear->WaitForPodNotFoundInNamespace replacement (parameter order not updated).

This is the same code with that problem fixed ("git diff 222f6550625..45770e38d5d"):

<details>

```patch
diff --git a/test/e2e/framework/network/utils.go b/test/e2e/framework/network/utils.go
index c0023eaf889..e4f733a57e6 100644
--- a/test/e2e/framework/network/utils.go
+++ b/test/e2e/framework/network/utils.go
@@ -888,7 +888,7 @@ func (config *NetworkingTestConfig) DeleteNetProxyPod(ctx context.Context) {
        framework.ExpectNoError(config.getPodClient().Delete(ctx, pod.Name, *metav1.NewDeleteOptions(0)))
        config.EndpointPods = config.EndpointPods[1:]
        // wait for pod being deleted.
-       err := e2epod.WaitForPodNotFoundInNamespace(ctx, config.f.ClientSet, config.Namespace, pod.Name, wait.ForeverTestTimeout)
+       err := e2epod.WaitForPodNotFoundInNamespace(ctx, config.f.ClientSet, pod.Name, config.Namespace, wait.ForeverTestTimeout)
        if err != nil {
                framework.Failf("Failed to delete %s pod: %v", pod.Name, err)
        }
diff --git a/test/e2e/framework/pod/pod_client.go b/test/e2e/framework/pod/pod_client.go
index 8c323d5a3e7..946587ae1df 100644
--- a/test/e2e/framework/pod/pod_client.go
+++ b/test/e2e/framework/pod/pod_client.go
@@ -181,7 +181,7 @@ func (c *PodClient) DeleteSync(ctx context.Context, name string, options metav1.
        if err != nil && !apierrors.IsNotFound(err) {
                framework.Failf("Failed to delete pod %q: %v", name, err)
        }
-       framework.ExpectNoError(WaitForPodNotFoundInNamespace(ctx, c.f.ClientSet, namespace, name, timeout), "wait for pod %q to disappear", name)
+       framework.ExpectNoError(WaitForPodNotFoundInNamespace(ctx, c.f.ClientSet, name, namespace, timeout), "wait for pod %q to disappear", name)
 }
 
 // mungeSpec apply test-suite specific transformations to the pod spec.
diff --git a/test/e2e/framework/volume/fixtures.go b/test/e2e/framework/volume/fixtures.go
index 3f33a5ec123..d8b62b5b0dc 100644
--- a/test/e2e/framework/volume/fixtures.go
+++ b/test/e2e/framework/volume/fixtures.go
@@ -463,7 +463,7 @@ func runVolumeTesterPod(ctx context.Context, client clientset.Interface, timeout
        }
        if err != nil {
                e2epod.DeletePodOrFail(ctx, client, clientPod.Namespace, clientPod.Name)
-               _ = e2epod.WaitForPodNotFoundInNamespace(ctx, client, clientPod.Namespace, clientPod.Name, timeouts.PodDelete)
+               _ = e2epod.WaitForPodNotFoundInNamespace(ctx, client, clientPod.Name, clientPod.Namespace, timeouts.PodDelete)
                return nil, err
        }
        return clientPod, nil
@@ -541,7 +541,7 @@ func testVolumeClient(ctx context.Context, f *framework.Framework, config TestCo
                // testVolumeClient might get used more than once per test, therefore
                // we have to clean up before returning.
                e2epod.DeletePodOrFail(ctx, f.ClientSet, clientPod.Namespace, clientPod.Name)
-               framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, clientPod.Namespace, clientPod.Name, timeouts.PodDelete))
+               framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, clientPod.Name, clientPod.Namespace, timeouts.PodDelete))
        }()
 
        testVolumeContent(f, clientPod, "", fsGroup, fsType, tests)
@@ -576,7 +576,7 @@ func InjectContent(ctx context.Context, f *framework.Framework, config TestConfi
                // This pod must get deleted before the function returns becaue the test relies on
                // the volume not being in use.
                e2epod.DeletePodOrFail(ctx, f.ClientSet, injectorPod.Namespace, injectorPod.Name)
-               framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, injectorPod.Namespace, injectorPod.Name, timeouts.PodDelete))
+               framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, injectorPod.Name, injectorPod.Namespace, timeouts.PodDelete))
        }()
 
        ginkgo.By("Writing text file contents in the container.")
diff --git a/test/e2e/kubectl/kubectl.go b/test/e2e/kubectl/kubectl.go
index a8c212bda5c..f5574e83cfd 100644
--- a/test/e2e/kubectl/kubectl.go
+++ b/test/e2e/kubectl/kubectl.go
@@ -734,7 +734,7 @@ metadata:
                                if !strings.Contains(ee.String(), "timed out") {
                                        framework.Failf("Missing expected 'timed out' error, got: %#v", ee)
                                }
-                               framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, ns, "failure-3", 2*v1.DefaultTerminationGracePeriodSeconds*time.Second))
+                               framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, "failure-3", ns, 2*v1.DefaultTerminationGracePeriodSeconds*time.Second))
                        })
 
                        ginkgo.It("[Slow] running a failing command with --leave-stdin-open", func(ctx context.Context) {
diff --git a/test/e2e/storage/pd.go b/test/e2e/storage/pd.go
index 755af5a08f1..8aa86b58a2c 100644
--- a/test/e2e/storage/pd.go
+++ b/test/e2e/storage/pd.go
@@ -195,7 +195,7 @@ var _ = utils.SIGDescribe("Pod Disks [Feature:StorageProvider]", func() {
                                        ginkgo.By("deleting host0Pod") // delete this pod before creating next pod
                                        framework.ExpectNoError(podClient.Delete(ctx, host0Pod.Name, podDelOpt), "Failed to delete host0Pod")
                                        framework.Logf("deleted host0Pod %q", host0Pod.Name)
-                                       e2epod.WaitForPodNotFoundInNamespace(ctx, cs, host0Pod.Namespace, host0Pod.Name, f.Timeouts.PodDelete)
+                                       e2epod.WaitForPodNotFoundInNamespace(ctx, cs, host0Pod.Name, host0Pod.Namespace, f.Timeouts.PodDelete)
                                        framework.Logf("deleted host0Pod %q disappeared", host0Pod.Name)
                                }
 
diff --git a/test/e2e/storage/testsuites/provisioning.go b/test/e2e/storage/testsuites/provisioning.go
index 4a5dc6e151b..87395546363 100644
--- a/test/e2e/storage/testsuites/provisioning.go
+++ b/test/e2e/storage/testsuites/provisioning.go
@@ -845,7 +845,7 @@ func (t StorageClassTest) TestBindingWaitForFirstConsumerMultiPVC(ctx context.Co
        framework.ExpectNoError(err)
        ginkgo.DeferCleanup(func(ctx context.Context) error {
                e2epod.DeletePodOrFail(ctx, t.Client, pod.Namespace, pod.Name)
-               return e2epod.WaitForPodNotFoundInNamespace(ctx, t.Client, pod.Namespace, pod.Name, t.Timeouts.PodDelete)
+               return e2epod.WaitForPodNotFoundInNamespace(ctx, t.Client, pod.Name, pod.Namespace, t.Timeouts.PodDelete)
        })
        if expectUnschedulable {
                // Verify that no claims are provisioned.
diff --git a/test/e2e/storage/ubernetes_lite_volumes.go b/test/e2e/storage/ubernetes_lite_volumes.go
index 8919a4898e4..252c1015502 100644
--- a/test/e2e/storage/ubernetes_lite_volumes.go
+++ b/test/e2e/storage/ubernetes_lite_volumes.go
@@ -100,7 +100,7 @@ func PodsUseStaticPVsOrFail(ctx context.Context, f *framework.Framework, podCoun
                        go func(config *staticPVTestConfig) {
                                defer ginkgo.GinkgoRecover()
                                defer wg.Done()
-                               err := e2epod.WaitForPodNotFoundInNamespace(ctx, c, ns, config.pod.Name, f.Timeouts.PodDelete)
+                               err := e2epod.WaitForPodNotFoundInNamespace(ctx, c, config.pod.Name, ns, f.Timeouts.PodDelete)
                                framework.ExpectNoError(err, "while waiting for pod to disappear")
                                errs := e2epv.PVPVCCleanup(ctx, c, ns, config.pv, config.pvc)
                                framework.ExpectNoError(utilerrors.NewAggregate(errs), "while cleaning up PVs and PVCs")
diff --git a/test/e2e/windows/density.go b/test/e2e/windows/density.go
index 76854b7da12..e0ab84ae10f 100644
--- a/test/e2e/windows/density.go
+++ b/test/e2e/windows/density.go
@@ -276,7 +276,7 @@ func deletePodsSync(ctx context.Context, f *framework.Framework, pods []*v1.Pod)
                        err := e2epod.NewPodClient(f).Delete(ctx, pod.ObjectMeta.Name, *metav1.NewDeleteOptions(30))
                        framework.ExpectNoError(err)
 
-                       err = e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, f.Namespace.Name, pod.ObjectMeta.Name, 10*time.Minute)
+                       err = e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, pod.ObjectMeta.Name, f.Namespace.Name, 10*time.Minute)
                        framework.ExpectNoError(err)
                }(pod)
        }
diff --git a/test/e2e_node/node_problem_detector_linux.go b/test/e2e_node/node_problem_detector_linux.go
index 904804cac03..39d25e9765c 100644
--- a/test/e2e_node/node_problem_detector_linux.go
+++ b/test/e2e_node/node_problem_detector_linux.go
@@ -433,7 +433,7 @@ current-context: local-context
                        ginkgo.By("Delete the node problem detector")
                        framework.ExpectNoError(e2epod.NewPodClient(f).Delete(ctx, name, *metav1.NewDeleteOptions(0)))
                        ginkgo.By("Wait for the node problem detector to disappear")
-                       gomega.Expect(e2epod.WaitForPodNotFoundInNamespace(ctx, c, ns, name, pollTimeout)).To(gomega.Succeed())
+                       gomega.Expect(e2epod.WaitForPodNotFoundInNamespace(ctx, c, name, ns, pollTimeout)).To(gomega.Succeed())
                        ginkgo.By("Delete the config map")
                        framework.ExpectNoError(c.CoreV1().ConfigMaps(ns).Delete(ctx, configName, metav1.DeleteOptions{}))
                        ginkgo.By("Clean up the events")
diff --git a/test/e2e_node/resource_collector.go b/test/e2e_node/resource_collector.go
index 5472213511a..60178b9d764 100644
--- a/test/e2e_node/resource_collector.go
+++ b/test/e2e_node/resource_collector.go
@@ -383,7 +383,7 @@ func deletePodsSync(ctx context.Context, f *framework.Framework, pods []*v1.Pod)
                                framework.Failf("Unexpected error trying to delete pod %s: %v", pod.Name, err)
                        }
 
-                       framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, f.Namespace.Name, pod.ObjectMeta.Name, 10*time.Minute))
+                       framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, pod.ObjectMeta.Name, f.Namespace.Name, 10*time.Minute))
                }()
        }
        wg.Wait()
```

</details>

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @pacoxu

pohly committed:
e2e framework: implement pod polling with gomega.Eventually

gomega.Eventually provides better progress reports: instead of filling up the
log with rather useless one-line messages that are not enough to to understand
the current state, it integrates with Gingko's progress reporting (SIGUSR1,
--poll-progress-after) and then dumps the same complete failure message as
after a timeout. That makes it possible to understand why progress isn't
getting made without having to wait for the timeout.

The other advantage is that the failure message for some unexpected pod state
becomes more readable: instead of encapsulating it as "observed object" inside
an error, it directly gets rendered by gomega.

pohly committed:
e2e framework: gomega assertions as errors

Calling gomega.Expect/Eventually/Consistently deep inside a helper call chain
has several challenges:
- the stack offset must be tracked correctly, otherwise the callstack
  for the failure starts at some helper code, which is often not informative
- augmenting the failure message with additional information from each
  caller implies that each caller must pass down a string and/or format
  string plus arguments

Both challenges can be solved by returning errors:
- the stacktrace is taken at that level where the error is
  treated as a failure instead of passing back an error, i.e.
  inside the It callback
- traditional error wrapping can add additional information, if
  desirable

What was missing was some easy way to generate an error via a gomega
assertion. The new infrastructure achieves that by mirroring the
Gomega/Assertion/AsyncAssertion interfaces with errors as return values instead
of calling a fail handler.

It is intentionally less flexible than the gomega APIs:
- A context must be passed to Eventually/Consistently as first
  parameter because that is needed for proper timeout handling.
- No additional text can be added to the failure through this
  API because error wrapping is meant to be used for this.
- No need to adjust the callstack offset because no backtrace
  is recorded when a failure occurs.

To avoid the useless "unexpected error" log message when passing back a gomega
failure, ExpectNoError gets extended to recognize such errors and then skips
the logging.

pohly committed:
e2e framework: include additional stack backtrace in failures

When a Gomega failure is converted to an error, the stack at the time when the
failure occurs may be useful: error wrapping provides some bread crumbs that
can be followed to determine where the failure really occurred, but error
wrapping may be missing or ambiguous.

To provide the additional information, a FailureError now includes a full stack
backtrace. The backtrace intentionally makes no attempt to exclude framework
functions besides the gomega support itself because helpers like
e2e/framework/pod may be relevant.

That backtrace is not included in the failure message for the sake of
brevity. Instead, it gets logged as part of the test's output.

pohly committed:
e2e: remove unused label filter from WaitForPodsRunningReady

None of the users of the functions passed anything other than nil or an empty
map and the implementation ignore the parameter - it seems like a candidate for
simplification.

pohly committed:
e2e framework: turn function into gomega.Matcher

The intention is to use this inside a helper function where the
corresponding Expect call is known.

pohly committed:
e2e pod: use gomega.Eventually in WaitForPodsRunningReady

The code becomes simpler (78 insertions, 91 deletions), easier to read (all
code entirely inside WaitForPodsRunningReady, no need to declare and later
overwrite variables) and possibly more correct (if all API calls failed,
the resulting error was ignored when allowedNotReadyPods > 0).

pohly committed:
e2e framework: support getting list of objects

This is similar to the previous support for getting a single object.

pohly committed:
e2e pod: use gomega.Eventually in WaitTimeoutForPodReadyInNamespace/WaitForPodCondition

These get converted together because they relied on FinalErr which now isn't
needed anymore.

pohly committed:
e2e: consolidate checking a pod list

WaitForPods is now a generic function which lists pods and then checks the pods
that it found against some provided condition. A parameter determines how many
pods must be found resp. match the condition for the check to succeed.

pohly committed:
e2e: replace WaitForPodToDisappear with WaitForPodNotFoundInNamespace

WaitForPodToDisappear was always called such that it listed all pods, which
made it less efficient than trying to get just the one pod it was checking for.

Being able to customize the poll interval in practice wasn't useful, therefore
it can be replaced with WaitForPodNotFoundInNamespace.

pohly committed:
e2e pod: use gomega.Eventually in WaitForPodNotFoundInNamespace

pohly committed:
e2e: consolidate pod response checking

This renames PodsResponding to WaitForPodsResponding for the sake of
consistency and adds a timeout parameter. That is necessary because some other
users of NewProxyResponseChecker used a much lower timeout (2min vs. 15min).

Besides simplifying some code, it also makes it easier to rewrite
ProxyResponseChecker because it only gets used in WaitForPodsResponding.

pohly committed:
e2e pod: convert ProxyResponseChecker into matcher

Instead of pod responses being printed to the log each time polling fails, we
get a consolidated failure message with all unexpected pod responses if (and
only if) the check times out or a progress report gets produced.

pohly committed:
e2e pod: use gomega.Eventually in WaitForRestartablePods

pohly committed:
e2e pod: remove dead code

pohly committed:
e2e storage: remove WaitForPodTerminatedInNamespace

Calling WaitForPodTerminatedInNamespace after testFlexVolume is useless because
the client pod that it waits for always gets deleted by testVolumeClient:

https://github.com/kubernetes/kubernetes/blob/0fcc3dbd55732c501d4948840f639e00a2fe33c5/test/e2e/framework/volume/fixtures.go#L541-L546

Worse, because WaitForPodTerminatedInNamespace treats "not found" as "must keep
polling", these two tests always kept waiting for 5 minutes:

    Kubernetes e2e suite: [It] [sig-storage] Flexvolumes should be mountable
    when non-attachable 	6m4s

The only reason why these tests passed is that WaitForPodTerminatedInNamespace
used to return the "not found" API error. That is not guaranteed and about to
change.

pohly committed:
e2e pod: unit test for pod status + API error

This covers new behavior in gomega.

pohly committed:
e2e: use error wrapping with %w

The recently introduced failure handling in ExpectNoError depends on error
wrapping: if an error prefix gets added with `fmt.Errorf("foo: %v", err)`, then
ExpectNoError cannot detect that the root cause is an assertion failure and
then will add another useless "unexpected error" prefix and will not dump the
additional failure information (currently the backtrace inside the E2E
framework).

Instead of manually deciding on a case-by-case basis where %w is needed, all
error wrapping was updated automatically with

    sed -i "s/fmt.Errorf\(.*\): '*\(%s\|%v\)'*\",\(.* err)\)/fmt.Errorf\1: %w\",\3/" $(git grep -l 'fmt.Errorf' test/e2e*)

This may be unnecessary in some cases, but it's not wrong.

aojea commented:
can we split this into more reviewable changes?

https://github.com/kubernetes/kubernetes/pull/115548/commits/45770e38d5d26181146acfaaba14fd20bbe9b84d

this commit seems can be done separately

pohly commented:
> Conflicting files
> test/e2e/framework/pod/wait.go
> test/e2e/framework/pod/wait_test.go 

This was a lie. All of the commits applied cleanly. Looks like "git merge" didn't like to re-apply the same commits. I had to rebase.

pohly commented:
> https://github.com/kubernetes/kubernetes/commit/45770e38d5d26181146acfaaba14fd20bbe9b84d
> this commit seems can be done separately

It can be merged separately, but it is needed for "wait for pods with gomega" to be complete.

pohly commented:
> can we split this into more reviewable changes?

I already split it into individual commits. Just look at it commit-by-commit instead of all files at once.

aojea commented:
sorry for the dumb questions, couldn't catch up with this, too many things :(

pohly commented:
/assign @xmcqueen

Can you perhaps review again? It's the exact same code as before, except for the additional swap of parameters in https://github.com/kubernetes/kubernetes/pull/115548#issuecomment-1419192535 (see PR description for that diff).

xmcqueen commented:
Yes, I'll review it again @pohly

pohly commented:
/retest

xmcqueen commented on test/e2e/framework/pod/wait.go:
```diff
@@ -637,140 +527,180 @@ func WaitForPodSuccessInNamespaceSlow(ctx context.Context, c clientset.Interface
-		options := metav1.ListOptions{LabelSelector: label.String()}
-		pods, err := c.CoreV1().Pods(ns).List(ctx, options)
+// PodsResponding waits for the pods to response.
+func WaitForPodsResponding(ctx context.Context, c clientset.Interface, ns string, controllerName string, wantName bool, timeout time.Duration, pods *v1.PodList) error {
```
This section seems to be not identical to the change in the referenced commit.  I see here the removal of `WaitForPodToDisappear` which matches the referenced commit, and I see the subsequent function below it `PodsResponding`, and in between I see the introduction of a new function `WaitForPodsResponding`. Is that per expectation? This new function seems to use the "namespace, name" sequence rather than the "name,namespace" introduced in the referenced commit.  Is this as intended?

xmcqueen replied:
I don't see a test case for the new function, but I do see the new function being used, so I expect this is as intended.  Leave this comment open for @pohly double-check.

xmcqueen replied:
Its in the original commit. close this comment

xmcqueen commented on test/e2e/framework/volume/fixtures.go:
```diff
@@ -464,7 +463,7 @@ func runVolumeTesterPod(ctx context.Context, client clientset.Interface, timeout
 	if err != nil {
 		e2epod.DeletePodOrFail(ctx, client, clientPod.Namespace, clientPod.Name)
-		_ = e2epod.WaitForPodToDisappear(ctx, client, clientPod.Namespace, clientPod.Name, labels.Everything(), framework.Poll, timeouts.PodDelete)
+		_ = e2epod.WaitForPodNotFoundInNamespace(ctx, client, clientPod.Name, clientPod.Namespace, timeouts.PodDelete)
```
noted x 3

xmcqueen commented on test/e2e/kubectl/kubectl.go:
```diff
@@ -411,7 +411,7 @@ var _ = SIGDescribe("Kubectl client", func() {
 			podYaml = commonutils.SubstituteImageName(string(readTestFileOrDie("pod-with-readiness-probe.yaml.in")))
 			e2ekubectl.RunKubectlOrDieInput(ns, podYaml, "create", "-f", "-")
-			framework.ExpectEqual(e2epod.CheckPodsRunningReady(ctx, c, ns, []string{simplePodName}, framework.PodStartTimeout), true)
+			framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(ctx, c, simplePodName, ns, framework.PodStartTimeout))
```
noted

xmcqueen commented on test/e2e/kubectl/kubectl.go:
```diff
@@ -1500,7 +1498,7 @@ metadata:
 			podYaml = commonutils.SubstituteImageName(string(readTestFileOrDie("pause-pod.yaml.in")))
 			e2ekubectl.RunKubectlOrDieInput(ns, podYaml, "create", "-f", "-")
-			framework.ExpectEqual(e2epod.CheckPodsRunningReady(ctx, c, ns, []string{pausePodName}, framework.PodStartTimeout), true)
+			framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(ctx, c, pausePodName, ns, framework.PodStartTimeout))
```
this looks right but is not in the referenced commit x 2 see below

xmcqueen replied:
in the original commit so close this comment

xmcqueen commented on test/e2e/storage/pd.go:
```diff
@@ -196,7 +195,7 @@ var _ = utils.SIGDescribe("Pod Disks [Feature:StorageProvider]", func() {
 					framework.ExpectNoError(podClient.Delete(ctx, host0Pod.Name, podDelOpt), "Failed to delete host0Pod")
 					framework.Logf("deleted host0Pod %q", host0Pod.Name)
-					e2epod.WaitForPodToDisappear(ctx, cs, host0Pod.Namespace, host0Pod.Name, labels.Everything(), framework.Poll, f.Timeouts.PodDelete)
+					e2epod.WaitForPodNotFoundInNamespace(ctx, cs, host0Pod.Name, host0Pod.Namespace, f.Timeouts.PodDelete)
```
noted

xmcqueen commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -846,7 +845,7 @@ func (t StorageClassTest) TestBindingWaitForFirstConsumerMultiPVC(ctx context.Co
 	ginkgo.DeferCleanup(func(ctx context.Context) error {
 		e2epod.DeletePodOrFail(ctx, t.Client, pod.Namespace, pod.Name)
-		return e2epod.WaitForPodToDisappear(ctx, t.Client, pod.Namespace, pod.Name, labels.Everything(), framework.Poll, t.Timeouts.PodDelete)
+		return e2epod.WaitForPodNotFoundInNamespace(ctx, t.Client, pod.Name, pod.Namespace, t.Timeouts.PodDelete)
```
noted

xmcqueen commented on test/e2e/storage/ubernetes_lite_volumes.go:
```diff
@@ -101,7 +100,7 @@ func PodsUseStaticPVsOrFail(ctx context.Context, f *framework.Framework, podCoun
 				defer ginkgo.GinkgoRecover()
 				defer wg.Done()
-				err := e2epod.WaitForPodToDisappear(ctx, c, ns, config.pod.Name, labels.Everything(), framework.Poll, f.Timeouts.PodDelete)
+				err := e2epod.WaitForPodNotFoundInNamespace(ctx, c, config.pod.Name, ns, f.Timeouts.PodDelete)
```
noted

xmcqueen commented on test/e2e/windows/density.go:
```diff
@@ -276,8 +276,7 @@ func deletePodsSync(ctx context.Context, f *framework.Framework, pods []*v1.Pod)
 
-			err = e2epod.WaitForPodToDisappear(ctx, f.ClientSet, f.Namespace.Name, pod.ObjectMeta.Name, labels.Everything(),
-				30*time.Second, 10*time.Minute)
+			err = e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, pod.ObjectMeta.Name, f.Namespace.Name, 10*time.Minute)
```
noted

xmcqueen commented on test/e2e_node/node_problem_detector_linux.go:
```diff
@@ -434,7 +433,7 @@ current-context: local-context
 			framework.ExpectNoError(e2epod.NewPodClient(f).Delete(ctx, name, *metav1.NewDeleteOptions(0)))
 			ginkgo.By("Wait for the node problem detector to disappear")
-			gomega.Expect(e2epod.WaitForPodToDisappear(ctx, c, ns, name, labels.Everything(), pollInterval, pollTimeout)).To(gomega.Succeed())
+			gomega.Expect(e2epod.WaitForPodNotFoundInNamespace(ctx, c, name, ns, pollTimeout)).To(gomega.Succeed())
```
noted

xmcqueen commented on test/e2e_node/resource_collector.go:
```diff
@@ -384,8 +383,7 @@ func deletePodsSync(ctx context.Context, f *framework.Framework, pods []*v1.Pod)
 
-			gomega.Expect(e2epod.WaitForPodToDisappear(ctx, f.ClientSet, f.Namespace.Name, pod.ObjectMeta.Name, labels.Everything(),
-				30*time.Second, 10*time.Minute)).NotTo(gomega.HaveOccurred())
+			framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, pod.ObjectMeta.Name, f.Namespace.Name, 10*time.Minute))
```
noted

xmcqueen commented:
Reviewed.  I see the new code from the referenced commit and it looks good.  The rest looks good, as before.

pacoxu commented:
/retest

bart0sh commented:
/triage accepted
/priority important-soon

bart0sh commented:
/retest

pacoxu reviewed (approved):
The problem with the reverted PR is mainly for WaitForPodNotFoundInNamespace.

/lgtm

/cc @aojea 
for double check
